### ISSUE TITLE
RobinsonToLK fails if paramodulation rewrites multiple positions

### ISSUE THREAD

**Issue by gebner** [gapt/gapt#452]
The paramodulation rule in the resolution calculus can rewrite multiple positions at once, while `Paramodulation{Left,Right}Rule` can only rewrite at a single position.

For example:
```scala
val Seq(a,b) = Seq("a","b") map {FOLConst(_)}
val p = FOLAtomConst("p", 2)
RobinsonToLK(
  Paramodulation(
    InputClause(Clause() :+ (a === b)), Suc(0),
    InputClause(Clause() :+ p(a,a)), Suc(0),
    p(b,b)
  )
)
```
fails with
```
at.logic.gapt.proofs.lkNew.LKRuleCreationException:
            Cannot create EqualityRightRule:
            Formulas p(a,a) and p(b,b) don't differ in exactly one position.
	at at.logic.gapt.proofs.lkNew.ConvenienceConstructor.LKRuleCreationException(lkNew.scala:1836)
	at at.logic.gapt.proofs.lkNew.EqualityRightRule$.apply(lkNew.scala:1603)
	at at.logic.gapt.proofs.lkNew.ParamodulationRightRule$.apply(macroRules.scala:1354)
	at at.logic.gapt.proofs.resolution.RobinsonToLK$$anonfun$at$logic$gapt$proofs$resolution$RobinsonToLK$$f$1$1.apply(RobinsonToLK.scala:94)
...
```

A similar problem occurs in definition elimination, when eliminating a definition increases the number of occurrences of a subterm.

So I guess we should change all of `Equality{Left,Right}MacroRule` and `Paramodulation{Left,Right}Rule` to simulate rewriting at multiple positions.  @loewenheim Any thoughts?

**Comment by quicquid** [comment]
With multiple replacements in a single inference it becomes significantly harder to check the rule. Also I was always wondering if the rule allows multiple replacements like P(a) -> P(f(a)) -> P(f(f(a))) with replacements a ->f(a) at 1.1, a -> f(a) at 1.1.1.

**Comment by gebner** [comment]
In the resolution calculus, we only support multiple *parallel* replacements.

IMO there are two issues here:

1. What is a resonable formalization of an equality rule?  (Ideally this should be consistent across the calculi.)
2. In what ways do we need to create them?

ad 1)  I can see a few possibilities here:
* Take a list of positions, and replace the subterm at each one of them with the right side of the equation.
* Take a single position, and replace the subterm there with the right side. (If you use this for paramodulation, you cannot easily simulate rewriting at a list of positions.  At least, not without a lot of factoring.)
* Take a lambda abstraction of a formula (or a formula with a free variable, whatever you prefer), and replace (the beta-reduced version of) `(\x. F) s` with `(\x. F) t` (I think this one is useful in HOL, otherwise you can't always apply equations of abstractions.)
* (Somewhat orthogonally:) Do we want a flag that specifies the direction of the equation (i.e. left-to-right, right-to-left), or a separate flip rule?

ad 2)
* For the prover9 import, I'm pretty sure we only need (first-order) rewriting at a single position.  (Otherwise we would have seen errors in the regression tests.)
* However for definition elimination, we do need rewriting at multiple positions (that's how I stumbled upon this bug), at least for non-right-linear rules.
* For RobinsonToLK, we need to simulate the resolution paramodulation rule in LK.

**Comment by gebner** [comment]
Fixed in commit 15a2645edcefa8dc191abe669c2d7c9f3216cea4
